### PR TITLE
bpo-40816 Add AsyncContextDecorator class

### PR DESCRIPTION
**Pull request by heckad**
https://bugs.python.org/issue40816

### CODE REVIEW

**Review by eamanu — CHANGES_REQUESTED** [review]
IMHO, we need test here, docs.. I thing this need more discussion?

**Comment by heckad** [comment]
Do you know you can write something like 
```
@contextmanager
def safe_space(is_on=True):
    token = IS_SAFE_SPACE_VALUE.set(is_on)
    try:
        yield
    finally:
        IS_SAFE_SPACE_VALUE.reset(token)
```

And use it as 

```
class MyClass:
    @safe_space()
    def safe_method(self):
        pass
```
but you can't write 
```
@asynccontextmanager
async def async_safe_space(is_on=True):
    token = IS_SAFE_SPACE_VALUE.set(is_on)
    try:
        yield
    finally:
        IS_SAFE_SPACE_VALUE.reset(token)
```
because you get the error `TypeError: '_AsyncGeneratorContextManager' object is not callable`. This pull request makes '_AsyncGeneratorContextManager' callable. This is an implementation of lost functionality.

**Comment by heckad** [comment]
@1st1?

**Comment by njsmith** [comment]
@belm0 asked me to look at this, but tbh I'm not real comfortable with adding new API surface to `contextlib` without input from the module owners, which I think is @ncoghlan and @1st1.

The idea seems straightforward enough. @eamanu is right thought that it definitely needs docs and tests.

**Comment by heckad** [comment]
@njsmith, I will add tests if this pull request is interesting. I was just waiting for a response from @1st1.

**Comment by belm0** [comment]
@heckad since core review attention is a scarce resource, I suggest having the PR ready to approve.  Since the tests and docs can mostly be copied from ContextDecorator, not too much work?

This functionality is needed-- let's assume the maintainers will agree.

**Comment by heckad** [comment]
@belm0 How to improve the documentation?

File: Lib/test/test_contextlib_async.py
Comment (belm0) [inline Lib/test/test_contextlib_async.py:282]: should it at least be called `test_decorator`?

But for the non-async case there are extensive tests for ContextDecorator.  They should be reviewed and copied where relevant.

https://github.com/python/cpython/blob/858a98fbe9446e4ec781b7f3caaf90505ce2ca23/Lib/test/test_contextlib.py#L395-L552

**Comment by belm0** [comment]
> @belm0 How to improve the documentation?

See context manager docs:

> contextmanager() uses ContextDecorator so the context managers it creates can be used as decorators as well as in with statements. When used as a decorator, a new generator instance is implicitly created on each function call (this allows the otherwise “one-shot” context managers created by contextmanager() to meet the requirement that context managers support multiple invocations in order to be used as decorators).

(it would also need "Changed in version X.X" annotation)

Also see the ContextDecorator docs.  Perhaps it's enough to say "Similar to ContextDecorator(), but works with async context managers" and give a short example.

https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager
https://docs.python.org/3/library/contextlib.html#contextlib.ContextDecorator

File: Lib/test/test_contextlib_async.py
Comment (heckad) [reply Lib/test/test_contextlib_async.py:282]: Yes, forgot this, thanks.

File: Lib/contextlib.py
Comment (1st1) [inline Lib/contextlib.py]:
```suggestion
    def _recreate_cm(self):
```

Comment (heckad) [reply Lib/contextlib.py]: Done

Comment (heckad) [reply Lib/contextlib.py]: Is the test written correctly? It turns out that we have called a function that does not exist because the test should have failed.

File: Lib/test/test_contextlib_async.py
Comment (1st1) [inline Lib/test/test_contextlib_async.py:283]: Can you also add `ncols = 0`

File: Lib/test/test_contextlib_async.py
Comment (1st1) [inline Lib/test/test_contextlib_async.py:293]: and `nonlocal ncols; ncols += 1`

File: Lib/test/test_contextlib_async.py
Comment (1st1) [inline Lib/test/test_contextlib_async.py:306]: And `self.assertEqual(ncols, 10)`

**Review by 1st1 — APPROVED** [review]

**Review by 1st1 — CHANGES_REQUESTED** [review]
Oh wait! The code/tests look great! But we also need to update the docs before we merge this.

**Comment by heckad** [comment]
Need to add a version. Which version of python will be added this feature?

File: Doc/library/contextlib.rst
Comment (1st1) [inline Doc/library/contextlib.rst:381]: Actually, can you just add an example to https://docs.python.org/3/library/contextlib.html#contextlib.asynccontextmanager. I don't think we need to document the class itself.

Comment (heckad) [reply Doc/library/contextlib.rst:381]: Maybe add a reference for more consistency. Something like `asynccontextmanager() uses AsyncContextDecorator so the context managers it creates can be used as decorators same as contextmanager.` And add ref from `AsyncContextDecorator` to `class:: AsyncContextManager`

**Comment by 1st1** [comment]
> Need to add a version. Which version of python will be added this feature?

3.10.

File: Doc/library/contextlib.rst
Comment (1st1) [reply Doc/library/contextlib.rst:381]: I wouldn't bother with anything but a 3 line example and saying it's possible starting with 3.10.

**Comment by heckad** [comment]
> > Need to add a version. Which version of python will be added this feature?
> 
> 3.10.

What about 3.9?

**Comment by 1st1** [comment]
We missed that train, unfortunately. 3.9 is frozen for new features.

File: Doc/library/contextlib.rst
Comment (1st1) [inline Doc/library/contextlib.rst]: `AsyncContextDecorator` is a low-level implementation detail. It doesn't need to be exposed because it will confuse a regular Python user.  Here's how I'd rewrite the doc:

```suggestion
   Context managers defined with :func:`asynccontextmanager` can be used
   either as decorators or with :keyword:`async with` statements::
   
     import time
   
     async def timeit():         
         now = time.monotonic()
         try:
             yield
         finally:
             print(f'it took {time.monotonic() - now}s to run')
             
      @timeit()
      async def main():
          # ... async code ...

   When used as a decorator, a new generator instance is implicitly created on
   each function call. This allows the otherwise "one-shot" context managers
   created by :func:`asynccontextmanager` to meet the requirement that context
   managers support multiple invocations in order to be used as decorators.

  .. versionchanged:: 3.10
     Async context managers created with :func:`asynccontextmanager` can
     be used as decorators.
```

File: Doc/library/contextlib.rst
Comment (belm0) [inline Doc/library/contextlib.rst:150]: Wrong version?

File: Doc/library/contextlib.rst
Comment (belm0) [reply Doc/library/contextlib.rst]: This language was copied from contextmanager docs, so the same argument could be applied there.  The docs of both managers should be consistent.

Comment (1st1) [reply Doc/library/contextlib.rst]: I'd say that the contextmanager docs should be updated to match my suggestion.

File: Doc/library/contextlib.rst
Comment (1st1) [reply Doc/library/contextlib.rst:150]: Why? This is a new feature and 3.9 is about to be released so it's frozen for new features. 3.10 is the one that will include this.

File: Doc/library/contextlib.rst
Comment (1st1) [reply Doc/library/contextlib.rst]: But there's no "rule" that the docs for different (albeit similar) APIs must look the same.

File: Doc/library/contextlib.rst
Comment (belm0) [reply Doc/library/contextlib.rst:150]: sorry, I misread it as 3.1

**Comment by heckad** [comment]
@1st1, all change was applied.

**Review by asvetlov — APPROVED** [review]

**Comment by asvetlov** [comment]
Thanks @heckad

**Pull request merged**
